### PR TITLE
[v1.18.x] prov/shm: change recv entry freestack into bufpool 

### PR DESCRIPTION
**Pull request by shijin-aws**
Currently smr_rx_entry is popped from a freestack of a fixed size (rx_size). However, both posted recv and unexpected message allocation will pop rx entry from the same stack, which makes it drained when the number of posted_recv + unexp_message >= rx_size.

This patch switch srx->recv_fs into srx->rx_pool as a buffer pool that can grow, and makes smr_rx_entry allocated from the rx pool

Note that this PR is opened directly to v1.18.x branch because shm's rx entry is already allocated from buffer pool in main branch, after using util srx.

### CODE REVIEW

**Review by lrbison — APPROVED** [review]

File: prov/shm/src/smr_ep.c
Comment (sunkuamzn) [inline prov/shm/src/smr_ep.c:1623]: Why 16 byte alignment?

**Review by sunkuamzn — APPROVED** [review]

**Comment by shijin-aws** [comment]
New revision added a test to check this fix

File: prov/shm/src/smr_ep.c
Comment (shijin-aws) [reply prov/shm/src/smr_ep.c:1623]: I followed how shm created the cmd_ctx bufpool https://github.com/ofiwg/libfabric/pull/8882/files, which has similar size with rx_entry.

**Review by aingerson — APPROVED** [review]

**Comment by shijin-aws** [comment]
bot:aws:retest

**Comment by shijin-aws** [comment]
Still the cuda-to-host fabtests timeout

```
server_command: ssh -n -o StrictHostKeyChecking=no -o ConnectTimeout=30 -o BatchMode=yes 172.31.42.42 'timeout 360 /bin/bash --login -c '"'"'/home/ec2-user/PortaFiducia/build/libraries/libfabric/pr9187-debug/install/fabtests/bin/fi_rma_bw -e rdm -o write -I 5 -v -p shm -s 172.31.42.42'"'"''

client_command: ssh -n -o StrictHostKeyChecking=no -o ConnectTimeout=30 -o BatchMode=yes 172.31.42.42 'timeout 360 /bin/bash --login -c '"'"'/home/ec2-user/PortaFiducia/build/libraries/libfabric/pr9187-debug/install/fabtests/bin/fi_rma_bw -e rdm -o write -I 5 -v -D cuda -p shm -s 172.31.42.42 172.31.42.42'"'"''
client_stdout:

client returncode: None
Client error: Client timed out
server_stdout:

server returncode: 124
```

will try to reproduce now

**Comment by shijin-aws** [comment]
Is ` continuous-integration/jenkins/pr-head Pending — This commit is being built ` a hang in tests or something else? @aingerson

**Comment by aingerson** [comment]
@shijin-aws All our clusters were powered off this weekend for maintenance. They should be back online now. I'll retrigger but they are unrelated as of right now! :)

**Comment by shijin-aws** [comment]
OK, I will wait for Intel CI before merging it

**Pull request merged**
